packages/flutter/lib/src/widgets/scroll_delegate.dart

```diff
@@ -0,0 +1,862 @@
+  ///   final List<Widget> _children = <Widget>[];
+  ///
+  ///   void someHandler() {
+  ///     setState(() {
+  ///       // The key here allows Flutter to reuse the underlying render
+  ///       // objects even if the children list is recreated.
+  ///       _children.add(ChildWidget(key: UniqueKey()));
+  ///     });
+  ///   }
+  ///
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     // Always create a new list of children as a Widget is immutable.
+  ///     return PageView(children: List<Widget>.of(_children));
+  ///   }
+  /// }
+  /// ```
+  final List<Widget> children;
+
+  /// A map to cache key to index lookup for children.
+  ///
+  /// _keyToIndex[null] is used as current index during the lazy loading process
+  /// in [_findChildIndex]. _keyToIndex should never be used for looking up null key.
+  final Map<Key?, int>? _keyToIndex;
+
+  bool get _isConstantInstance => _keyToIndex == null;
+
+  int? _findChildIndex(Key key) {
+    if (_isConstantInstance) {
+      return null;
+    }
+    // Lazily fill the [_keyToIndex].
+    if (!_keyToIndex!.containsKey(key)) {
+      int index = _keyToIndex![null]!;
+      while (index < children.length) {
+        final Widget child = children[index];
+        if (child.key != null) {
+          _keyToIndex![child.key] = index;
+        }
+        if (child.key == key) {
+          // Record current index for next function call.
+          _keyToIndex![null] = index + 1;
+          return index;
+        }
+        index += 1;
+      }
+      _keyToIndex![null] = index;
+    } else {
+      return _keyToIndex![key];
+    }
+    return null;
+  }
+
+  @override
+  int? findIndexByKey(Key key) {
+    final Key childKey;
+    if (key is _SaltedValueKey) {
+      final _SaltedValueKey saltedValueKey = key;
+      childKey = saltedValueKey.value;
+    } else {
+      childKey = key;
+    }
+    return _findChildIndex(childKey);
+  }
+
+  @override
+  Widget? build(BuildContext context, int index) {
+    if (index < 0 || index >= children.length) {
+      return null;
+    }
+    Widget child = children[index];
+    final Key? key = child.key != null? _SaltedValueKey(child.key!) : null;
+    if (addRepaintBoundaries) {
+      child = RepaintBoundary(child: child);
+    }
+    if (addSemanticIndexes) {
+      final int? semanticIndex = semanticIndexCallback(child, index);
+      if (semanticIndex != null) {
+        child = IndexedSemantics(index: semanticIndex + semanticIndexOffset, child: child);
+      }
+    }
+    if (addAutomaticKeepAlives) {
+      child = AutomaticKeepAlive(child: _SelectionKeepAlive(child: child));
+    }
+
+    return KeyedSubtree(key: key, child: child);
+  }
+
+  @override
+  int? get estimatedChildCount => children.length;
+
+  @override
+  bool shouldRebuild(covariant SliverChildListDelegate oldDelegate) {
+    return children != oldDelegate.children;
+  }
+}
+class _SelectionKeepAlive extends StatefulWidget {
```

Comment:
nit: add a blank line before this class?

Reply:
```suggestion

class _SelectionKeepAlive extends StatefulWidget {
```